### DISCUSSION TITLE
[Feature Request] - Auto dark mode

### DISCUSSION THREAD

**Discussion by balintbabics** [gethomepage/homepage#1892] [Feature Requests]
### Description

Hello there,

I couldn't find the required format for feature request here, so apologies if it is missing some formal formatting.

**The idea**:

I'm really enjoying using homepage in my home lab, however I find it a bit odd that a feature-packed service like this is missing auto dark mode option (could be fixed time based or "dynamic").

**The suggestion**:

So my suggestion will be a settings option for theme, to be auto, which could be configured to be "dynamic" (location - long/lat based nighttime/daytime | sunset/sunrise) or fixed scheduled time based.

**The problem**:

If someone not uses the default light/dark theme, it should be somehow configurable to which theme should be used during daytime, and which one during nighttime.

**Comment by balintbabics** [comment]
Oh wow, I thought this idea would make a bit bigger impression on users. Well, I was wrong.

**Comment by sanello84** [comment]
I would recommend it follow the browser's theme or "based on device settings" and maybe add a third toggle to the bottom of the page where the current toggle resides. 

https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

**Comment by benphelps** [comment]
This functionality existed in the past and was removed due to the FOUC (flash of unstyled content) effect it would cause.

Any setting that makes part of the display dynamic, untimitally leads to the page style being rendered slower, and people expect near 0ms load times from homepage, with absolutely no FOUC.

It caused more people to complain about the flash, than it made people happy.

**Reply by Tywele** [reply]
Did the effect occur even when the option was disabled?

**Comment by vember31** [comment]
I was able to use some custom javascript to automatically change the browser localStorage entry for theme-mode when the theme for the browser changes from light to dark mode or vice versa.  Ultimately when you click the toggle in Homepage, it updates that localStorage theme-mode entry.  

My browser's theme (I use Vivaldi) is tied to the OS theme, which is managed by AutoDarkMode (and AutoDarkMode sets the OS theme based on windows night light. windows night light turns on based on dynamic time of sunrise/sunset).  So, the browser theme goes to dark mode at night, and light mode during the day.

If you put the following javascript in your custom.js file that lives at /app/config/custom.js, it will listen for changes in your browser's theme and change from light to dark (or dark to light) in response. To see the changed setting, Homepage must reload, so this script reloads Homepage as well if a theme change is noticed.  Yes, it's a slight delay for reload, but it goes quickly and doesn't bother me personally :)

Hopefully this is useful for someone.

```
// Function to set theme-mode in localStorage and reload the page if necessary
function setThemeMode() {
    // Check if the browser is in dark mode
    const darkMode = window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches;

    // Determine the new theme-mode
    const newThemeMode = darkMode ? 'dark' : 'light';

    // Check the current theme-mode in localStorage
    const currentThemeMode = localStorage.getItem('theme-mode');

    // If the theme-mode has changed, update it and reload the page
    if (currentThemeMode !== newThemeMode) {
        localStorage.setItem('theme-mode', newThemeMode);
        console.log('Theme mode changed to:', newThemeMode);
        location.reload();
    }
}

// Listen for changes in the browser's theme and update theme-mode accordingly
window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', setThemeMode);

// Initial check to set theme-mode when the script loads
setThemeMode();
```

**Comment by F0ofoon** [comment]
Just found a simpler solution that works fine (flash still happens, but better to have that and an auto theme switcher IMO).

Code has been generated with AI.

```js
// Function to update theme class on #page_wrapper
function updateTheme() {
    const wrapper = document.getElementById('page_wrapper');
    if (!wrapper) return;

    if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        wrapper.className = 'relative dark scheme-dark theme-neutral';
    } else {
        wrapper.className = 'relative light scheme-light theme-neutral';
    }
}

// Update theme on page load
updateTheme();

// Listen to system color scheme changes
window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', updateTheme);

```

**Comment by AuroraTheOtter** [comment]
Would love to have a auto theme setting

**Comment by tjmarkham** [comment]
I was looking for something like this and didn't find anything that worked. Here's an updated version of what was posted before.

This can be pasted to the top of your `config/custom.js` file:

```javascript
// auto dark mode

function updateTheme(darkModeTest) {
  const htmlElement = document.querySelector('html');

  if(! htmlElement) {
    return;
  }

  const themeClassList = htmlElement.classList;

  if (darkModeTest.matches) {
    themeClassList.add('scheme-dark', 'dark');
    themeClassList.remove('scheme-light', 'light');
  } else {
    themeClassList.add('scheme-light', 'light');
    themeClassList.remove('scheme-dark', 'dark');
  }
}

const darkModeQuery = window.matchMedia('(prefers-color-scheme: dark)');

updateTheme(darkModeQuery);

darkModeQuery.addEventListener('change', updateTheme);
```
